### PR TITLE
minor fixup for LDLT_factor::inverse

### PR DESCRIPTION
**Pull request by jpritikin**

### CODE REVIEW

**Comment by syclik** [comment]
Jenkins, ok to test.

**Comment by bob-carpenter** [comment]
What's up with signed/unsigned here?  `size_t` is unsigned, whereas `Eigen::DenseIndex` is signed (defaults to `std::ptrdiff_t` in the more recent versions of Eigen).  We'll get comparison warnings if you try to compare them in the code.

Any int value in a Stan program should be signed.

File: stan/math/prim/mat/fun/LDLT_factor.hpp [hunk 1]
```diff
@@ -119,10 +119,10 @@ namespace stan {
...
-      inline size_t rows() const { return N_; }
-      inline size_t cols() const { return N_; }
+      inline Eigen::DenseIndex rows() const { return N_; }
+      inline Eigen::DenseIndex cols() const { return N_; }
 
-      typedef size_t size_type;
+      typedef Eigen::DenseIndex size_type;
       typedef double value_type;
 
       size_t N_;
```
Comment (bob-carpenter) [inline stan/math/prim/mat/fun/LDLT_factor.hpp]: I'm guessing `N_` here should be redeclared as being `Eigen::DenseIndex` (aka `size_type` in this class scope).  Otherwise, we're going to get warnings about losing precision or sign information.

**Comment by jpritikin** [comment]
@bob-carpenter What's up with signed/unsigned here?

In OpenMx, we `#define EIGEN_DEFAULT_DENSE_INDEX_TYPE int`

**Comment by bob-carpenter** [comment]
int and ptrdiff_t are signed.  size_t is unsigned.
When you try to compare or assign signed to unsigned
or vice versa, the compiler gripes.

Most of our code's using int for Eigen indexing,
which is also wrong.  We really should be doing it everywhere
the typedef the way you did it.  

We no longer override any of the Eigen built-in types
because they led to order dependencies on the includes which
caused build bugs that were hard to track down.

- Bob

File: stan/math/prim/mat/fun/LDLT_factor.hpp [hunk 2]
```diff
@@ -94,8 +94,8 @@ namespace stan {
         return _ldltP->vectorD().array().log().sum();
       }
 
-      inline void inverse(Eigen::Matrix<T, R, C> &invA) const {
-        invA.setIdentity(N_);
+      template <typename MB> inline void inverse(Eigen::MatrixBase<MB> &invA) const {
```
Comment (syclik) [inline stan/math/prim/mat/fun/LDLT_factor.hpp]: @jpritikin, it's failing cpplint.
Can you split this into two lines? (you're on your fork, which makes it harder for me to fix directly) the natural place is after the template argument.

**Comment by syclik** [comment]
Jenkins, retest this please.

**Comment by syclik** [comment]
Jenkins, retest this please.

File: stan/math/prim/mat/fun/LDLT_factor.hpp (hunk 1 above)
Comment (syclik) [inline stan/math/prim/mat/fun/LDLT_factor.hpp]: @jpritikin, can you update line 128?

Comment (jpritikin) [inline stan/math/prim/mat/fun/LDLT_factor.hpp]: Oh! Sorry for the delay. I realize what Bob had pointed out.

Comment (syclik) [inline stan/math/prim/mat/fun/LDLT_factor.hpp]: No problem. (I would have fixed it myself if it was easier to create a pull
request to your fork, but that's not so straightforward.)

**Comment by syclik** [comment]
@jpritikin, sorry there hasn't been any activity on this pull request. I know it's been brought up in multiple places, but the only hesitation I have with this is I don't know what types we should test with to verify that `MatrixBase<MB>` covers enough.

I understand that Eigen claims that it does, but I thought there were a couple of edge cases where blocks or some other slicing operator didn't work, but I could be mistaken. What types do use use within OpenMX?

**Comment by bob-carpenter** [comment]
It might help to start with the goal here.  Is it to
save needless copies by changing the return value of
our functions to Eigen template expressions?  If so,
we just need to make sure everything we output (which we
control) matches every argument position of the right type
in our other functions.

I believe block() was where we ran into the problem that
the return value doesn't extend their MatrixBase, but was
assignable to Matrix.  So their doc isn't telling the whole
story up front when they talk about MatrixBase:

http://eigen.tuxfamily.org/dox/TopicFunctionTakingEigenTypes.html

But if block() doesn't work, that means our own block function
can't return a template expression;  but we could just force
the copy ourselves by returning Matrix<T,...> and that
would be assignable to MatrixBase. 

But you see they go on to talk explicitly about block()
when they talk about Ref<Derived>, which can handle block()
template expressions, so maybe that's what we should be using.

I find their doc for all this incredibly confusing.

- Bob

**Comment by syclik** [comment]
Thanks for the much needed, clearer explanation.

**Comment by jpritikin** [comment]
> Is it to save needless copies by changing the return value of our functions to Eigen template expressions?

I think the proposal focuses on changing the argument types more than the return type. If the arguments take a MatrixBase<T> then we can pass Eigen::Map< .. > in and everything works without an extra copy. In terms of testing, I don't understand the concern. The compiler should be able to check for type compatibility at compile time. I don't think this change introduces any runtime testing burden.

**Comment by syclik** [comment]
tl;dr: please include additional tests for Eigen expression template types that could be used. 

Re: testing. Yes, the compiler will catch it eventually, but not in the test for LDLT_factor, if there is a problem. I would like to know the minimal set of things we would need to test for input arguments and verify that this function works for all of them. 

If you think it's just Matrix, Block, and Map, then please include that in the tests. As Bob mentioned, we tried at some point and one of the template expressions weren't of type MatrixBase and the compiler let us know. We know that Matrix will work, but I am uncertain about MatrixBase.

If you're able to show that it works for the expression templates and what tests you used to show that, I'd be happy with a more general code base.

**Comment by jpritikin** [comment]
> I would like to know the minimal set of things we would need to test for input arguments and verify that this function works for all of them.

You make the burden for contribution too high. By switching to `MatrixBase`, the API is becoming more general. Everything that was supported before continues to work. I am not keen on writing tests for types that may never see use in actual code. That just seems like a waste of time. Testing might make sense if there was an alternative to `MatrixBase<T>`, but I am not aware of another choice. There is only one choice to make the API more flexible.

**Comment by syclik** [comment]
Hi, all that I'm asking for are additional tests for:
- `Eigen::Block`
- `Eigen::Map`
Unless you think there's an additional expression template types that
OpenMX would need. (I've already outlined what I thought we needed in Stan:
`Matrix` and `Block`.)

Regarding making the burden for contribution too high: I am trying to
prevent this particular change causing a compiler error for a user down the
line. It's possible by generalizing this, we could have expressions in the
Stan language that get translated to C++ fine, but get caught by the C++
compiler. That's not acceptable behavior. That edge case is not something
we have tested for exhaustively given we haven't had the need with
`Eigen::Matrix`. So, I feel it's a completely justified burden of
contribution. Just prove that this change is safe and it gets in.

Thanks.

**Comment by betanalpha** [comment]
I’m with Dan — the tests just need to instantiate the template as 
is expected, either in use cases within Stan or for OpenMX.  They
don’t need to test everything, just where we need them to work.

**Comment by jpritikin** [comment]
Well, I frankly fail to see the point of this, but it's pretty easy to add the tests.

**Comment by syclik** [comment]
@jpritikin, the test failure isn't your fault -- it's due to your branch (specifically the commit hash) not being somewhere on stan-dev/math. We can fix that later.

Now that I'm looking at this pull request, I'm a bit confused:
1. The only place you added `MatrixBase` was to the `inverse()` method. Was that intentional? I would prefer everything to be consistent.
2. The additional test you added didn't touch the `inverse()` method, so it wasn't actually testing your pull request.

I dug up the short example that caused Bob and I so much trouble. I can share that with you if you'd like to see why we're not so trusting of what the doc in Eigen claims.

**Comment by jpritikin** [comment]
> I dug up the short example that caused Bob and I so much trouble. I can share that with you if you'd like to see why we're not so trusting of what the doc in Eigen claims.

Sure please

**Comment by syclik** [comment]
@jpritikin, I forwarded an email to you. There were a couple responses on the Eigen list and it was also forwarded to stan-dev.

Products of matrices don't behave well. In this particular case (inverse in place), it's user error, but in general, we can't blanket replace all instances with MatrixBase because if a product can be provided with as an argument, we could have problems if we access individual elements. I'm not sure what happens if we use it as an expression.

**Comment by jpritikin** [comment]
> Was that intentional?

No, I'm just dumb.

> I would prefer everything to be consistent.

OK, but that is going to be a much larger patch to convert all of LDLT_factor.

**Comment by syclik** [comment]
Is there any reason you started with LDLT_factor? Now that I've looked at it again, I think it should be rewritten. If there is no particular use case, maybe start with something simpler? We can reason through all the use cases without having to worry about this particular class that has weird stated I'll properties.

**Comment by jpritikin** [comment]
> Is there any reason you started with LDLT_factor?

I can't remember. I just look for open pull requests and tried to make progress here.

I think the one I'm really interested in is multi_normal_log.hpp and then pull #40. Maybe that's why I started with LDLT_factor, because these will depend on it.

What about LDLT_factor needs to be rewritten?

**Comment by bob-carpenter** [comment]
> On Mar 2, 2016, at 8:39 AM, Joshua Pritikin <notifications@github.com> wrote:
> 
> I would like to know the minimal set of things we would need to test for input arguments and verify that this function works for all of them.
> 
> You make the burden for contribution too high. By switching to MatrixBase, the API is becoming more general.

> Everything that was supported before continues to work.

It explicitly disallows the block expression templates, which
we use internally to avoid copies.

> I am not keen on writing tests for types that may never see use in actual code.

Nobody is.  That's why we need to clearly define how whatever
is changing is going to get called and test for that.

> That just seems like a waste of time. Testing might make sense if there was an alternative to MatrixBase<T>, but I am not aware of another choice. There is only one choice to make the API more flexible.

Ref<Derived> is an alternative, but it needs to have the Derived
class specified or type inference won't work, so that's also not general
enough.   Only Matrix<T,R,C> seems to support the most general level of
castability and assignability.

- Bob

**Comment by syclik** [comment]
@jpritikin, ok to close this pull request?
